pkg/registries/registries.go

```diff
@@ -35,81 +36,162 @@ func ScopeIsNestedInsideScope(subScope, superScope string) bool {
+		sortedRepos = sortedRepos[:len(sortedRepos)-1]
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	return sortedRepos, nil
+}
 
+type mergedMirrorSet struct {
+	source             string
+	mirrors            []string
+	mirrorSourcePolicy apicfgv1.MirrorSourcePolicy
+}
+
+// mergedMirrorSets converts the set of mirrors to slice of mergedMirrorSet
+func mergedMirrorSets(sets *mirrorSets) ([]mergedMirrorSet, error) {
 	// Sort the sets of mirrors by Source to ensure deterministic output
 	sources := []string{}
-	for key := range disjointSets {
+	for key := range sets.disjointSets {
 		sources = append(sources, key)
 	}
+	// collects the mirror sources and sorted in increasing order
 	sort.Strings(sources)
 	// Convert the sets of mirrors
-	res := []apioperatorsv1alpha1.RepositoryDigestMirrors{}
+	res := []mergedMirrorSet{}
 	for _, source := range sources {
-		ds := disjointSets[source]
-		topoGraph := newTopoGraph()
-		for _, set := range *ds {
-			for i := 0; i+1 < len(set.Mirrors); i++ {
-				topoGraph.AddEdge(set.Mirrors[i], set.Mirrors[i+1])
-			}
-			sourceInGraph := false
-			for _, m := range set.Mirrors {
-				if m == source {
-					sourceInGraph = true
-					break
-				}
-			}
-			if !sourceInGraph {
-				// The build of mirrorSets guarantees len(set.Mirrors) > 0.
-				topoGraph.AddEdge(set.Mirrors[len(set.Mirrors)-1], source)
-			}
-			// Every node in topoGraph, including source, is implicitly added by topoGraph.AddEdge (every mirror set contains at least one non-source mirror,
-			// so there are no unconnected nodes that we would have to add separately from the edges).
-		}
-		sortedRepos, err := topoGraph.Sorted()
+		mirrors, err := sets.mergedMirrors(source)
 		if err != nil {
 			return nil, err
 		}
-		if sortedRepos[len(sortedRepos)-1] == source {
-			// We don't need to explicitly include source in the list, it will be automatically tried last per the semantics of sysregistriesv2. Mirrors.
-			sortedRepos = sortedRepos[:len(sortedRepos)-1]
+		item := mergedMirrorSet{
+			source:  source,
+			mirrors: mirrors,
 		}
-		res = append(res, apioperatorsv1alpha1.RepositoryDigestMirrors{
-			Source:  source,
-			Mirrors: sortedRepos,
-		})
+		if sets.mirrorBlockSource[source] {
+			item.mirrorSourcePolicy = apicfgv1.NeverContactSource
+		}
+		res = append(res, item)
 	}
 	return res, nil
 }
 
+// mergedTagMirrorSets processes itmsRules and returns a set of mergedMirrorSet, one for each Source value,
+// ordered consistently with the preference order of the individual entries (if possible)
+// E.g. given mirror sets (B, C) and (A, B), it will combine them into a single (A, B, C) set.
+func mergedTagMirrorSets(itmsRules []*apicfgv1.ImageTagMirrorSet) ([]mergedMirrorSet, error) {
+	tagMirrorSets := newMirrorSets()
+	for _, itms := range itmsRules {
+		for i := range itms.Spec.ImageTagMirrors {
+			set := itms.Spec.ImageTagMirrors[i]
```

Comment:
Absolutely non-blocking:
```go
for _, set := range … { /* use set */ }
```
would be one variable and one line shorter. (In all three instances.)